### DISCUSSION TITLE
React Query 사용 이유

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#93] [프론트 기술 질문 공간]
요즘 개발자들은 React Query를 왜 사용하나요?
React Query를 사용한 경험이 있다면 왜 사용했나요?

**Comment by michoball** [comment]
전형적으로 웹 프론트 프레임워크에서 사용하는 전역 상태관리 툴들은 프론트에서만 사용되는 상태 값들에 맞춰져 있습니다. 
그래서 서버에서 가져오는 데이터를 다룰때는 부수효과를 주어 데이터를 fetching 한 후 상태관리 툴에 주입해 사용하게 되는데요.
이런 방식을 사용하게 되면 프론트에서 사용되고 있는 서버 값이 실제 서버값과 다른 상태를 가질 수 있게 됩니다. 

 서버에서의 온전한 값을 사용하지 못한다는 단점이 발생하게 됩니다.  그래서 이를 위해 임의로 서버 데이터를 주기적으로 refetching하게 하는 코드를 추가적으로 작성해야하는데요.
이러한 단점을 보완한 것이 react-query 입니다.  react-query 를 사용하면 위의 문제를 해결하기 위한 코드를 작성할 필요가없고, 데이터를 refetching하는 시간을 조정할 수 있고 알아서 오래된 데이터를 다시 fetch하게 할 수도 있는 등 서버 상태 값을 다루기 위한 많은 작업을 수월하게 해주도록 합니다.  

 react-query를 사용해 본 입장에서 서버에서 데이터를 가져오는 부분의 코드량이 매우 줄어든 것 뿐만 아니라 로직이 좀 더 명확해지고 더 잘 이해가 되는 느낌을 받아서 계속 react-query를 사용하고자 하고 있습니다.

**Comment by WooGie911** [comment]
React Query는 서버 상태 관리를 위한 라이브러리입니다. 

React는 단방향 통신 형식으로 서버로부터 데이터를 요청하고 받는 작업을 할 때 대부분 리덕스와 리덕스 thunk 혹은 saga를 사용하는데, 
React Query를 사용하면 Boilerplate code를 대폭 줄일 수 있으며 상태관리에 대한 다양한 옵션을 제공하여 간편하게 적용할 수 있습니다. 

또한 선언적으로 프로그래밍 할 수 있어 코드를 절약할 수 있으며, 동일한 API 요청이 여러 번 호출될 경우 한 번만 실행해줍니다. 

그리고 데이터의 상태가 stale 하다고 판단되면 적절한 시점에 업데이트 해주며, 전역 상태와 서버 상태를 분리할 수 있다는 장점을 가지고 있습니다. 

게다가 React Query의 useQuery와 useMataion을 통해 CRUD 요청도 할 수 있습니다. 

만약 전역 상태관리가 필요한 경우 비교적 가벼운 리코일이나 contextAPI를 함께 활용하는 것도 좋습니다.

**Comment by javascriptbiggosoo** [comment]
TanStack Query는 React 애플리케이션에서 데이터를 더 쉽게 가져오고 관리하기 위해 만들어진 도구입니다.

일반적으로 React에서 데이터를 가져오기 위해 우리는 `useEffect`를 사용하는데, 이를 사용하면 데이터를 불러오는 동안 발생할 수 있는 여러 상황들을 개발자가 직접 처리해야 합니다. 
예를 들어, 데이터를 불러오는 동안 로딩 상태를 관리하거나, 데이터가 제대로 불러와지지 않았을 때의 오류 상태를 관리하거나, 페이지를 나갈 때 데이터를 불러오는 요청을 취소하는 등의 작업들이 필요합니다.

TanStack Query는 데이터를 불러오고 관리하는 과정을 간편하게 만들어줍니다. 
예를 들어, 데이터를 자동으로 캐시하므로 불필요한 네트워크 요청을 줄일 수 있습니다. 
또한 오류 발생 시 자동으로 재시도하고, 쿼리를 관리하며, 데이터를 업데이트하는 등의 기능을 제공합니다.
